pilot/pkg/security/authn/v1beta1/policy_applier.go

```diff
@@ -256,23 +265,45 @@ func convertToEnvoyJwtConfig(jwtRules []*v1beta1.JWTRule) *envoy_jwt.JwtAuthenti
 		}
 		provider.FromParams = jwtRule.FromParams
 
-		jwtPubKey := jwtRule.Jwks
-		if jwtPubKey == "" {
-			var err error
-			jwtPubKey, err = model.GetJwtKeyResolver().GetPublicKey(jwtRule.JwksUri)
-			if err != nil {
-				log.Errorf("Failed to fetch jwt public key from %q: %s", jwtRule.JwksUri, err)
-				// This is a temporary workaround to reject a request with JWT token by using a fake jwks when istiod failed to fetch it.
-				// TODO(xulingqing): Find a better way to reject the request without using the fake jwks.
-				jwtPubKey = createFakeJwks(jwtRule.JwksUri)
+		if features.EnableRemoteJwks && jwtRule.JwksUri != "" {
```

Comment:
I think this may break if a user has a `Sidecar` restricting scope. Do we have a plan for the future of this `EnableRemoteJwks` setting? We are trying to avoid permanent off feature flags

Reply:
Yes. The plan is to remove the EnableRemoteJwks once it is stabilized. If we can't find the cluster with Sidecar, we will fallback to old mode and will fail.

Reply:
>If we can't find the cluster with Sidecar, we will fallback to old mode and will fail.

I don't think the code is doing this today, right?

I am worried this will break people if its enabled. For example, they may have network/authz policy allowing access only from istiod

Reply:
It is doing. If the cluster is not found, it falls back to local jwks

```
if err == nil && len(cluster) > 0 {
				// This is a case of URI pointing to mesh cluster. Setup Remote Jwks and let Envoy fetch the key.
				provider.JwksSourceSpecifier = &envoy_jwt.JwtProvider_RemoteJwks{
					RemoteJwks: &envoy_jwt.RemoteJwks{
						HttpUri: &core.HttpUri{
							Uri: jwtRule.JwksUri,
							HttpUpstreamType: &core.HttpUri_Cluster{
								Cluster: cluster,
							},
							Timeout: &duration.Duration{Seconds: 5}, // TODO: Make this configurable.
						},
						CacheDuration: &duration.Duration{Seconds: 5 * 60}, // TODO: Make this configurable if needed.
					},
				}
			} else {
				provider.JwksSourceSpecifier = buildLocalJwks(jwtRule.JwksUri, "")
			}
			
```
and it fails to fetch the public key and fails.

Reply:
The cluster lookup is not taking sidecar scope into account as far as I can tell

Reply:
I see your point. The same logic is used by extAuthz as well. Should we add in the release notes that this cluster needs to be added in the sidecar if it is enabled similar to how we used to specify mixer?

Reply:
I don't think we agreed to remove `EnableRemoteJwks` even when this is stable. This is a new feature that changes to use Envoy to fetch jwksUri instead of Istiod. That's an option that should be configured by user, and the default would be Istiod as it's the existing behavior since the very beginning.

And in order to promote this feature to a more stable state (experimental -> alpha), we will at least need a doc describing the API change that controls the behavior, might be as simple as a flag in meshconfig or just keeping it an environmental variable.

Addressing the sidecar scope is another improvement to this feature, but should not be a blocking issue, could just be documented as a limitation of the feature.

Reply:
> I don't think we agreed to remove EnableRemoteJwks even when this is stable

What I meant was if it is working well and we detect that the jwksUri is a mesh cluster - we can remove that flag right? That is the way user is configuring that he wants to use remoteJwks. If it fails find the cluster, it will any fallback to local jwksUri.?

Reply:
We need to stop being too dependent on a specific behavior/implementation. 

Most env variables are by definition temporary - if this feature works well and has users we will either promote it to the API or make it default behavior if it makes sense. 

Like we did with SDS, where the impl moved from node to pod to istiod, which is now the default.

( I'm not sure if this particular feature should be default or not - we'll need to see )